### PR TITLE
AArch64: Implement ArrayCHKEvaluator()

### PR DESCRIPTION
**Pull request by sahithiponangi**
Issue: #6438

Signed-off-by: Siri Sahithi Ponangi <sahithi.ponangi@unb.ca>

### CODE REVIEW

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -3339,6 +3339,163 @@ J9::ARM64::TreeEvaluator::ArrayStoreCHKEvaluator(TR::Node *node, TR::CodeGenerat
...
+      gcPoint->ARM64NeedsGCMap(cg, 0x0);
+      snippet->gcMap().setGCRegisterMask(0x0);
+      }
+
+   cg->stopUsingRegister(tmp1Reg);
+   cg->stopUsingRegister(tmp2Reg);
+
+   cg->decReferenceCount(node->getFirstChild());
+   cg->decReferenceCount(node->getSecondChild());
+   return (NULL);
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:3489]: You don't need the parentheses in this line.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -3339,6 +3339,163 @@ J9::ARM64::TreeEvaluator::ArrayStoreCHKEvaluator(TR::Node *node, TR::CodeGenerat
...
+   // Same array, we are done.
+   //
+   generateCompareInstruction(cg, node, obj1Reg, obj2Reg, true);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, doneLabel, TR::CC_EQ);
+
+   // If we know nothing about either object, test object1 first. It has to be an array.
+   //
+   if (!node->isArrayChkPrimitiveArray1() && !node->isArrayChkReferenceArray1() && !node->isArrayChkPrimitiveArray2() && !node->isArrayChkReferenceArray2())
+      {
+
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:3375]: You don't need the empty line here.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -3339,6 +3339,163 @@ J9::ARM64::TreeEvaluator::ArrayStoreCHKEvaluator(TR::Node *node, TR::CodeGenerat
...
+      }
+   else
+      {
+      // We have to take care of the un-equal class situation: both of them must be of reference array
+      generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, doneLabel, TR::CC_EQ);
+
+      // Object1 must be of reference component type, otherwise throw exception
+      if (!node->isArrayChkReferenceArray1())
+         {
+
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:3419]: You don't need the empty line here.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 4]
```diff
@@ -3339,6 +3339,163 @@ J9::ARM64::TreeEvaluator::ArrayStoreCHKEvaluator(TR::Node *node, TR::CodeGenerat
...
+            cg->addSnippet(snippet);
+            }
+         else
+            generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, snippetLabel, TR::CC_NE);
+         }
+
+      // Object2 must be of reference component type array, otherwise throw exception
+      if (!node->isArrayChkReferenceArray2())
+         {
+
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:3448]: You don't need the empty line here.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 5]
```diff
@@ -3339,6 +3339,163 @@ J9::ARM64::TreeEvaluator::ArrayStoreCHKEvaluator(TR::Node *node, TR::CodeGenerat
...
+   cg->stopUsingRegister(tmp1Reg);
+   cg->stopUsingRegister(tmp2Reg);
+
+   cg->decReferenceCount(node->getFirstChild());
+   cg->decReferenceCount(node->getSecondChild());
+   return (NULL);
+   }
+
+
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:3492]: Please remove one of the empty lines here.

**Comment by knn-k** [comment]
@0xdaryl Do you have any comments on this PR? Thanks.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 6]
```diff
@@ -3339,6 +3339,159 @@ J9::ARM64::TreeEvaluator::ArrayStoreCHKEvaluator(TR::Node *node, TR::CodeGenerat
...
 
+static TR::Register *
+VMarrayCheckEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR::Register *obj1Reg = cg->evaluate(node->getFirstChild());
+   TR::Register *obj2Reg = cg->evaluate(node->getSecondChild());
+   TR::Register *tmp1Reg = cg->allocateRegister();
+   TR::Register *tmp2Reg = cg->allocateRegister();
+
+   TR::LabelSymbol *doneLabel, *snippetLabel;
```
Comment (0xdaryl) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:3350]: Move these  down to where they are first defined below.

**Review by 0xdaryl — COMMENTED** [review]
I think the title of the commit should simply be "AArch64: Implelemt ArrayCHKEvaluator()" and you get rid of the body of the commit except for the "Issue: #6438".

**Comment by 0xdaryl** [comment]
Jenkins test sanity alinux64 jdk11

**Comment by 0xdaryl** [comment]
jit_jitt_array_compress_0 and jit_jitt_array_compress_1 failures are known.

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**